sampling/rate_limited/lib/opentelemetry/sampling/rate_limited/sampler.rb

```diff
@@ -0,0 +1,63 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module Sampling
+    module RateLimited
+      class Sampler
+        def initialize(credits_per_second)
+          @credits_per_second = credits_per_second
+          @balance = 0
+          @last_tick = Time.now.to_i
+          @lock = Mutex.new
+        end
+
+        # @param [String] trace_id
+        # @param [OpenTelemetry::Context] parent_context
+        # @param [Enumerable<Link>] links
+        # @param [String] name
+        # @param [Symbol] kind
+        # @param [Hash<String, Object>] attributes
+        # @return [OpenTelemetry::SDK::Trace::Samplers::Result] The sampling result.
+        def should_sample?(trace_id:, parent_context:, links:, name:, kind:, attributes:)
+          if can_spend?
+            OpenTelemetry::SDK::Trace::Samplers::Result.new(
+              decision: OpenTelemetry::SDK::Trace::Samplers::Decision::RECORD_AND_SAMPLE,
+              tracestate: OpenTelemetry::Trace.current_span(parent_context).context.tracestate
+            )
+          else
+            OpenTelemetry::SDK::Trace::Samplers::Result.new(
+              decision: OpenTelemetry::SDK::Trace::Samplers::Decision::DROP,
+              tracestate: OpenTelemetry::Trace.current_span(parent_context).context.tracestate
+            )
+          end
+        end
+
+        private
+
+        # @return [Boolean]
+        def can_spend?
+          return false if @credits_per_second <= 0
+
+          @lock.synchronize do
```

Comment:
The sampler is set on the tracer provider so you'll be locking every time you create a span, this may not be appropriate depending on how much span volume your service is handling.

Reply:
Do you suggest being optimistic and not locking at all?

Reply:
I don't think we have a suggestion as much as we're doing is raising concerns about the performance implications of adding another lock around span creation.

I'm not sure how much latency this introduces for applications that generate thousands of spans in a single transaction.

Reply:
Since default thread counts are fairly low in most server implementations (in Puma it's `5`, for instance for MRI). We can also provide a configuration option for that